contracts/roles.js

```diff
@@ -0,0 +1,917 @@
+    }
+    if (res.errors) {
+      api.debug(`Error paying out roles of ${quantity} ${symbol} to ${account} (TXID ${api.transactionId}): \n${res.errors}`);
+      return false;
+    }
+    return true;
+  }
+  return false;
+}
+
+async function checkPendingCandidates(inst, params) {
+  const random = api.random();
+  const blockDate = new Date(`${api.hiveBlockTimestamp}.000Z`);
+  const upInst = JSON.parse(JSON.stringify(inst));
+  const voteTokenObj = await api.db.findOneInTable('tokens', 'tokens', { symbol: inst.voteToken });
+  const voteTokenMinValue = api.BigNumber(1)
+    .dividedBy(api.BigNumber(10).pow(voteTokenObj.precision));
+  const instTickTime = api.BigNumber(blockDate.getTime())
+    .minus(params.instanceTickHours * 3600 * 1000).toNumber();
+
+  let rolesProcessed = 0;
+  const pendingRoles = await api.db.find('roles',
+    {
+      instanceId: inst.id,
+      active: true,
+      'tokenBalances.0': { $exists: true },
+      lastTickTime: { $lte: instTickTime },
+    },
+    params.maxRolesPerBlock,
+    0,
+    [{ index: 'byLastTickTime', descending: false }, { index: '_id', descending: false }]);
+
+  for (let i = 0; i < pendingRoles.length; i += 1) {
+    const role = pendingRoles[i];
+    const funded = [];
+    const payTokens = role.tokenBalances.filter(t => api.BigNumber(t.quantity).gt(0));
+    const totalSlots = api.BigNumber(role.mainSlots).plus(role.backupSlots).toNumber();
+    const roleTickTime = api.BigNumber(blockDate.getTime())
+      .minus(role.tickHours * 3600 * 1000).toNumber();
+
+    if (role.lastTickTime <= roleTickTime) {
+      if (payTokens.length > 0) {
+        let offset = 0;
+        let candidates = await api.db.find('candidates',
+          {
+            roleId: role._id,
+            active: true,
+            approvalWeight: { $gt: { $numberDecimal: api.BigNumber(role.voteThreshold) } },
+          },
+          params.processQueryLimit,
+          offset,
+          [{ index: 'byApprovalWeight', descending: true }, { index: '_id', descending: false }]);
+
+        let accWeight = 0;
+        let backupWeight = null;
+        do {
+          for (let j = 0; j < candidates.length; j += 1) {
+            const candidate = candidates[j];
+            if (funded.length >= role.mainSlots && backupWeight === null) {
+              backupWeight = api.BigNumber(accWeight)
+                .plus(voteTokenMinValue)
+                .plus(api.BigNumber(role.totalApprovalWeight.$numberDecimal)
+                  .minus(accWeight)
+                  .times(random))
+                .toFixed(voteTokenObj.precision, api.BigNumber.ROUND_HALF_UP);
+            }
+
+            accWeight = api.BigNumber(accWeight)
+              .plus(candidate.approvalWeight.$numberDecimal)
+              .toFixed(voteTokenObj.precision, api.BigNumber.ROUND_HALF_UP);
+
+            if (candidate.active === true) {
+              if (funded.length < role.mainSlots || api.BigNumber(backupWeight).lte(accWeight)) {
```

Comment:
The backup mechanism seems a bit weird, or works only when numBackup = 1. If there are more than 1 backups requested, and it generates a single number, it'll just pick two backups together without re-rolling the die?  I think this might be similar to what witness contract does (Which is also wrong)

Reply:
Makes sense since it was copied from witness contract. Going to rework.